### PR TITLE
fix(database): drop fk before dropping index

### PR DESCRIPTION
**Pull request by innerdvations**
### What does it do?

Changes the order of operations in `syncSchema`. In some databases (like mysql) it is impossible to drop an index on a foreign key unless you first drop the fk. The previous order of removal was: index -> fk -> column, I've changed it to fk -> column -> index

### Why is it needed?

see above

more specifically, it also prevents the v4->v5 migration on mysql, because the long version of shortened db identifiers for fks cannot be dropped

### How to test it?

NOTE: there's currently an issue with document id migrations on mysql as well preventing startup, which will prevent you testing this as a v4->v5 migration directly, but otherwise doing a v4->v5 migrations would be the best way to test this since that's what this PR is really aiming to solve

- Using mysql 8
- create your own fk index in the db that strapi won't recognize and will try to delete on startup (or try to migrate from Strapi 4 to Strapi 5)
- start strapi
  - without this PR, it will fail to start because of the error described in https://github.com/strapi/strapi/issues/20020
  - with this PR, the fk will be dropped and Strapi will start fine

Given the comments below, also check some other cases that might trigger a reverse problem. Try restarting strapi after the following cases:

- remove relation (foreign key) from a schema
- add foreign keys to a schema
- rename a relation in a schema
- change the type of a relation to something else while keeping the same name
- swap the names of two attributes, relation and non-relation (but if there's a problem confirm if it exists on v5/main or develop as well, I think there are already potentially some bugs around that area)

Manually tested:

- v4 -> v5 migration
- dropping relations in the CTB
- dropping components in the CTB
- editing relations in the CTB

with

- [x] mysql
- [x] sqlite
- [x] postgres
- [x] mariadb

### Related issue(s)/PR(s)

Fixes https://github.com/strapi/strapi/issues/20020

DX-1374

### CODE REVIEW

**Review by alexandrebodin — COMMENTED** [review]
I'm a bit surprised, Does this work everywhere else ? I seem to remember I had to drop the index on foreign keys before dropping them to avoid errors in the past 🤔 

Otherwise code looks fine

**Comment by innerdvations** [comment]
> I'm a bit surprised, Does this work everywhere else ? I seem to remember I had to drop the index on foreign keys before dropping them to avoid errors in the past 🤔
> 
> Otherwise code looks fine

Oh, that's odd, I hope there's not some dialectal difference where postgres requires it one way and mysql requires it the other. I will add some additional notes on QA on this.

**Review by christiancp100 — DISMISSED** [review]
QAed it and found this error renaming and changing the relations content type for the following content type. Previously was Address and changed to Country: 
image
image

File: packages/core/database/src/schema/builder.ts [hunk 1]
```diff
@@ -273,6 +264,29 @@ const createHelpers = (db: Database) => {
         dropColumn(tableBuilder, removedColumn);
       }
 
+      // for mysql only, dropForeignKey also removes the index, so don't drop it twice
+      const isMySQL = db.config.connection.client === 'mysql';
+      const ignoreForeignKeyNames = isMySQL
+        ? [
+            ...table.foreignKeys.removed.map((fk) => fk.name),
```
Comment (alexandrebodin) [inline packages/core/database/src/schema/builder.ts:271]: Are we sure the fk and idx have the same names 🤔  don't we do _fk or _idx or sth like that ?

Comment (innerdvations) [reply packages/core/database/src/schema/builder.ts:271]: I might have something wrong here but from what I can tell:

There's a foreign key index, and a foreign key constraint. When we create the foreign key, we just give it the index name and they both use it.

[When we create them, we use the same name for both](https://github.com/strapi/strapi/blob/922ac347bb655c5d2cfd5f70ecf727f8199372fe/packages/core/database/src/schema/schema.ts#L94).

For postgres and sqlite, `dropForeignKey` drops the constraint, but not the index. For mysql, `dropForeignKey` drops both.

Comment (alexandrebodin) [reply packages/core/database/src/schema/builder.ts:271]: 👍

**Review by alexandrebodin — APPROVED** [review]
Code looks good, I didn't QA

**Comment by innerdvations** [comment]
> I did need to try starting the app twice after this change on the first start I got this error

I wonder if that's an issue upgrading from 5.0.0-beta.1 instead of from 4.x. I've tried all 4 supported databases now going from 4.x -> this PR and also tried modifying and dropping relations, and didn't have any problem.

**Pull request merged**